File: 232_Implement_Queue_using_Stacks.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/implement-queue-using-stacks/
# Implement the following operations of a queue using stacks.
# push(x) -- Push element x to the back of queue.
# pop() -- Removes the element from in front of queue.
# peek() -- Get the front element.
# empty() -- Return whether the queue is empty.
# You may assume that all operations are valid (e.g., no pop or peek operations will be called on an empty queue).

# One stack contains the iteams as they are added. Another stack is used to reverse the order during the pop operation.
# The top of the stack is recorded for O(1) peek.
# Time - O(1) for init, push, peek and empty. O(n) for pop.
# Space - O(n)

class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stack = []
        self.reversed = []
        self.top = None

    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: void
        """
        if not self.stack:
            self.top = x
        self.stack.append(x)

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        if not self.reversed:
            while self.stack:
                self.reversed.append(self.stack.pop())
        return self.reversed.pop()

    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        if self.reversed:
            return self.reversed[-1]
        return self.top

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return not self.stack and not self.reversed

File: java/232_Implement_Queue_using_Stacks.java
/*
https://leetcode.com/problems/implement-queue-using-stacks/
Implement the following operations of a queue using stacks.
push(x) -- Push element x to the back of queue.
pop() -- Removes the element from in front of queue.
peek() -- Get the front element.
empty() -- Return whether the queue is empty.

Use stackA to push and stackB to pop.  If stackB is empty then transfer all of stackA to stackB (inverts order).
Time - O(1) to push, O(n) (best case is O(1)) to pop and peek.
Space - O(n)
*/

class MyQueue {

    private Stack<Integer> stackA = new Stack<Integer>();
    private Stack<Integer> stackB = new Stack<Integer>();

    // Push element x to the back of queue.
    public void push(int x) {
        stackA.push(x);
    }

    // Removes the element from in front of queue.
    public void pop() {
        if (stackB.isEmpty())
            while (!stackA.isEmpty())
                stackB.push(stackA.pop());
        int x = stackB.pop();
    }

    // Get the front element.
    public int peek() {
        if (stackB.isEmpty())
            while (!stackA.isEmpty())
                stackB.push(stackA.pop());
        return stackB.peek();
    }

    // Return whether the queue is empty.
    public boolean empty() {
        return stackA.isEmpty() && stackB.isEmpty();
    }
}
